Tools Panel: only show header ➕ icon when all items are optional

ramonjd opened the pull request:
## Description

➕ ➕ ➕ ➕ ➕ ➕ ➕ ➕ 

This change checks for default ToolsPanel items and, where they exist, disables the plus icon.

The plus icon should only appear where there are optional menu items only, and only when no optional item is selected.

The argument to limit the plus icon in this way is that there is the potential for it to cause confusion as to its purpose, particularly in the presence of a minus sign (where the control has a value).

<img width="267" alt="Screen Shot 2022-01-27 at 1 20 59 pm" src="https://user-images.githubusercontent.com/6458278/151280333-e8182b74-c198-4667-9171-116f8368d7c9.png">

See discussion over at https://github.com/WordPress/gutenberg/issues/38219#issuecomment-1022751087 for context

➕ ➕ ➕ ➕ ➕ ➕ ➕ ➕ 

## Testing Instructions

Pick a block that has some form of block supports with default entries for example the List Block `block.json` has Typography support.

Fire up the editor and check that the ellipses icon for the corresponding panel is always, regardless of which item you select or deselect. 

<img width="400" alt="Screen Shot 2022-01-27 at 12 52 23 pm" src="https://user-images.githubusercontent.com/6458278/151278574-c78d3d96-4518-4836-a185-9f8ad68b9075.png">

Now, remove the `"__experimentalDefaultControls"` object:

```js
	"supports": {
		"anchor": true,
		"className": false,
		"typography": {
			"fontSize": true,
			"__experimentalFontFamily": true,
			"lineHeight": true,
			"__experimentalFontStyle": true,
			"__experimentalFontWeight": true,
			"__experimentalLetterSpacing": true,
			"__experimentalTextTransform": true,
			"__experimentalDefaultControls": {    // <-- Remove this object
				"fontSize": true
			}
		},
```

Check that the plus icon is displayed when there are no optional controls selected.

<img width="400" alt="Screen Shot 2022-01-27 at 12 55 09 pm" src="https://user-images.githubusercontent.com/6458278/151278591-e722ffec-15f8-44d5-a2ce-c8eff3771de6.png">

Blocks that inject controls into the ToolsPanel, e.g., minimum height for the Cover Block, or drop cap for the Paragraph block should work similarly depending on whether the ToolsPanel Item `isShownByDefault` prop is true or false.

So if it's `true` the plus icon will not show since there's a default. If it's `false` it means the fill is optional, so, if there are no other defaults, a plus icon will display.

<img width="286" alt="Screen Shot 2022-01-27 at 1 16 13 pm" src="https://user-images.githubusercontent.com/6458278/151279920-6acea400-f724-4b3b-bdec-439ecf73e8c6.png">

Tests shall pass! 

`npm run test-unit packages/components/src/tools-panel/test/index.js`

Ensure the relevant story is still... relevant. 

` npm run storybook:dev`


## Types of changes
Design tool icon change.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

ramonjd committed:
This change checks for default ToolsPanel items and, where they exist, disables the plus icon.
The plus icon should only appear where there are optional menu items only, and only when no optional item is selected.
Adding a test to reflect this change.

ramonjd committed:
Whoops. Forgot to remove the false check.

aaronrobertshaw commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -48,6 +48,13 @@ const generateMenuItems = ( {
+) =>
+	obj &&
+	Object.keys( obj ).length === 0 &&
+	Object.getPrototypeOf( obj ) === Object.prototype;
```
Given the typing here for `ToolsPanelMenuItems` do we need the additional check for the object prototype?

glendaviesnz replied:
I was thinking the same thing, seems like the check to ensure we are dealing with an object via the prototype is redundant?

ramonjd replied:
> do we need the additional check for the object prototype?

No, we do not. Good call.

aaronrobertshaw commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -167,24 +174,24 @@ export function useToolsPanel(
-				optionalItems.length > 0 &&
-				! optionalItems.some( ( [ , isSelected ] ) => isSelected );
+		if (
+			isMenuItemTypeEmpty( menuItems?.default ) &&
```
Now we've added `menuIitems.default` as a dependency for this `useEffect` we should probably add it to the hooks dependency array, change the dependency to only `menuItems`, or determine this separately and pass only the boolean result to the hook.

aaronrobertshaw reviewed (commented):
Thanks for putting this together so quick @ramonjd 🚀 

This tests as advertised for me:
✅ ToolsPanel unit tests pass
✅ New unit tests LGTM
✅ Vertical ellipsis is shown whenever the panel contains default controls or optional control is toggled on
✅ Plus icon is only shown when the panel contains all optional controls and all are hidden

I left a couple of small nitpick comments.

In addition, I think it would be good to update the `ToolsPanel` storybook example relating to the plus icon use to reflect this PR's changes.

For example, adding a "knob" to the story to toggle on or off inclusion of a default control so the conditions under which the plus icon should be shown can more easily be explored.

<details>
<summary><strong>Possible update to the ToolsPanel story</strong></summary>

```diff
diff --git a/packages/components/src/tools-panel/stories/index.js b/packages/components/src/tools-panel/stories/index.js
index f3afeb1190..f36076ee65 100644
--- a/packages/components/src/tools-panel/stories/index.js
+++ b/packages/components/src/tools-panel/stories/index.js
@@ -2,6 +2,7 @@
  * External dependencies
  */
 import styled from '@emotion/styled';
+import { boolean } from '@storybook/addon-knobs';
 
 /**
  * WordPress dependencies
@@ -23,6 +24,9 @@ import { createSlotFill, Provider as SlotFillProvider } from '../../slot-fill';
 export default {
 	title: 'Components (Experimental)/ToolsPanel',
 	component: ToolsPanel,
+	parameters: {
+		knobs: { disable: false },
+	},
 };
 
 export const _default = () => {
@@ -138,19 +142,38 @@ export const WithNonToolsPanelItems = () => {
 export const WithOptionalItemsPlusIcon = () => {
 	const [ height, setHeight ] = useState();
 	const [ width, setWidth ] = useState();
+	const [ minWidth, setMinWidth ] = useState();
 
 	const resetAll = () => {
 		setHeight( undefined );
 		setWidth( undefined );
+		setMinWidth( undefined );
 	};
 
+	const includeDefaultControls = boolean( 'includeDefaultControls', false );
+
 	return (
 		<PanelWrapperView>
 			<Panel>
 				<ToolsPanel
 					label="Tools Panel (optional items only)"
 					resetAll={ resetAll }
+					key={ includeDefaultControls }
 				>
+					{ includeDefaultControls && (
+						<SingleColumnItem
+							hasValue={ () => !! minWidth }
+							label="Minimum width"
+							onDeselect={ () => setMinWidth( undefined ) }
+							isShownByDefault={ true }
+						>
+							<UnitControl
+								label="Minimum width"
+								value={ minWidth }
+								onChange={ ( next ) => setMinWidth( next ) }
+							/>
+						</SingleColumnItem>
+					) }
 					<SingleColumnItem
 						hasValue={ () => !! width }
 						label="Width"

```
</details>

Also, we'll need an entry in the components package changelog. 🙂 

Thanks again for this one!

glendaviesnz commented:
This worked as advertised for me, and I think the UX makes more sense with this change.

ramonjd committed:
Updating story to include knob that toggles the default status of the min-width panel
updating isEmpty utility function to not check of object type. TS does that.
Updating useEffect dep array.

ramonjd commented:
Thanks for the reviews folks! And for the nudge about Storybook knobs @aaronrobertshaw (and for taking the time to make a patch)

![Jan-27-2022 16-41-48](https://user-images.githubusercontent.com/6458278/151298971-3a0b968f-6030-4d91-9c2e-e4f966dfe26e.gif)

aaronrobertshaw committed:
Add changelog

jasmussen commented:
Nice, incredibly fast work! Works as advertised:
![paragraph](https://user-images.githubusercontent.com/1204802/151309597-78009961-a6eb-4a26-b41e-1895d01fbc27.gif)

It's a small change, but for the purposes of clearing values from default controls, this feels like it might help. Additional tweaks we might explore, completely separate from this PR and if necessary, could be a conditional "clear all" icon that shows up next to the ellipsis.

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.ts:
```diff
@@ -48,6 +48,10 @@ const generateMenuItems = ( {
 };
 
+const isMenuItemTypeEmpty = (
+	obj: ToolsPanelMenuItems[ ToolsPanelMenuItemKey ]
```
If obj can be `undefined`, then this type should be updated to reflect that

```suggestion
	obj?: ToolsPanelMenuItems[ ToolsPanelMenuItemKey ]
```

ciampo commented on packages/components/src/tools-panel/stories/index.js:
```diff
@@ -138,19 +142,36 @@ export const WithNonToolsPanelItems = () => {
 				<ToolsPanel
 					label="Tools Panel (optional items only)"
 					resetAll={ resetAll }
+					key={ includeDefaultControls }
```
Is `includeDefaultControls` used as a `key` to force `ToolsPanel` to re-render?

ramonjd replied:
Yes. Unfortunately without it, the panel doesn't refresh properly. At first I thought it pointed to a bug in the source code, but the storybook and the Gutenberg app are behaving differently here. 🤷

ciampo reviewed (commented):
Thank you @ramonjd for working on this!

@aaronrobertshaw's review already included most of the comments that I would have made 🎉 

I just left a couple of minor comments, otherwise the changes in this PR LGTM :)

---

As a follow-up (with not rush at all), we should refactor this Storybook example as the `knobs` addon is [deprecated](https://github.com/storybookjs/storybook/discussions/15060). We should look into converting the examples to the new `controls` approach — here's a couple of references:

- [Controls docs](https://storybook.js.org/docs/react/essentials/controls)
- [a component in gutenberg using Controls in its storybook example](https://github.com/WordPress/gutenberg/blob/9faa63ec01c7db4e4d57823af2db5c54bb6d6660/packages/components/src/toggle-group-control/stories/index.js)

ramonjd committed:
Update packages/components/src/tools-panel/tools-panel/hook.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

ramonjd commented:
> As a follow-up (with not rush at all), we should refactor this Storybook example as the knobs addon is deprecated. We should look into converting the examples to the new controls approach — here's a couple of references:

Great to know, thanks @ciampo It's a good opportunity to read up on them. Thanks!!

ramonjd commented:
Migration to controls over at https://github.com/WordPress/gutenberg/pull/38308
